lib/tapioca/gem/listeners/mixins.rb

```diff
@@ -62,6 +74,40 @@ def add_mixins(tree, mods, mixin_type)
             end
         end
 
+        sig do
+          params(
+            location: String,
+          ).returns(T::Boolean)
+        end
+        def mixed_in_by_gem?(location)
+          @pipeline.gem.contains_path?(location)
+        end
+
+        sig do
+          params(
+            constant: Module,
+            mixin: Module,
+            mixin_type: Runtime::Trackers::Mixin::Type
+          ).returns(T.nilable(String))
+        end
+        def mixin_location(constant, mixin, mixin_type)
+          T.must(Runtime::Trackers::Mixin.constants_with_mixin(mixin)[mixin_type])[constant]
+        end
+
+        sig do
+          params(
+            constant: Module,
+            mixin: Module,
+          ).returns(T.nilable(String))
+        end
+        def singleton_class_mixin_location(constant, mixin)
+          singleton_class = singleton_class_of(constant)
+          mixin_locations = Runtime::Trackers::Mixin.constants_with_mixin(mixin)
+
+          T.must(mixin_locations[Runtime::Trackers::Mixin::Type::Include])[singleton_class] ||
+            T.must(mixin_locations[Runtime::Trackers::Mixin::Type::Prepend])[singleton_class]
```

Comment:
Again, `dig` would be much nicer:
```suggestion
          mixin_locations.dig(Runtime::Trackers::Mixin::Type::Include, singleton_class) ||
            mixin_locations.dig(Runtime::Trackers::Mixin::Type::Prepend, singleton_class)
```